### PR TITLE
Add env test and widget

### PR DESCRIPTION
**Pull request by anxolin**
This PR adds:

## ✅  Unit testing 
- Adds test integration using JEST
- Adds Typescript support for the tests and some basic config

## 🤩 Github Actions: Test + Lint 
- Make sure each PR now passes the tests and lint rules

See TEST https://github.com/cowprotocol/cow-fi/actions/runs/6788445276/job/18453501066?pr=146
and LINT https://github.com/cowprotocol/cow-fi/actions/runs/6788445276/job/18453500755?pr=146

## 🕵️ Environment detection
It will detect, based on the domain in which domain are we

Similar to CoW Swap logic, but with less env: no barn, no ENS, no staging

## 🧪 Tests for environment detection
image

## ~~Adapts the name and environment of the widget~~
- ~~Adds the environment of the app~~
- ~~Sets a new name~~

EDIT: I did the environment detection because i thought I had to pass it to the widget. But the env is only used to construct the baseURL of the iframe. Never mind, at least now we have the env detection, GH Actions, and unit tests in this repo

## Change name
https://dev.explorer.cow.fi/orders/0x342014318a7cee35d3d5e1f57705bcd5b308bb597524ebe281e9d58b9a5f31e279063d9173c09887d536924e2f6eadbabac099f5654a840a?tab=overview

image

### CODE REVIEW

File: util/environment.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+const DEFAULT_ENVIRONMENTS_REGEX: Record<EnvironmentName, string> = {
+  local: '^(:?localhost:\\d{2,5}|(?:127|192)(?:\\.[0-9]{1,3}){3})',
+  pr: '^cowfi-git-[\\w\\d-]+cowswap\\.vercel\\.app',
+  development: '^(develop.cow.fi|swap-develop.vercel.app)',
+  production: '^cow.fi$',
```
Comment (alfetopito) [inline util/environment.ts:5]: If we want to be complete, there are more prod domains to add 
image

Although it's probably overkill

File: util/environment.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+const DEFAULT_ENVIRONMENTS_REGEX: Record<EnvironmentName, string> = {
+  local: '^(:?localhost:\\d{2,5}|(?:127|192)(?:\\.[0-9]{1,3}){3})',
+  pr: '^cowfi-git-[\\w\\d-]+cowswap\\.vercel\\.app',
+  development: '^(develop.cow.fi|swap-develop.vercel.app)',
```
Comment (alfetopito) [inline util/environment.ts:4]: `swap-develop` is for CoW Swap: 
image

Only develop.cow.fi is for dev AFAICT

image

File: util/environment.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  production: '^cow.fi$',
+}
+
+function getRegex(env: EnvironmentName) {
+  const regex = process.env[`REACT_APP_DOMAIN_REGEX_${env.toUpperCase()}`] || DEFAULT_ENVIRONMENTS_REGEX[env]
```
Comment (alfetopito) [inline util/environment.ts:9]: Given the problem you recently solved with dynamic env vars with the RPC url, could this cause the same problem?

Comment (anxolin) [reply util/environment.ts:9]: Good point!

Comment (anxolin) [reply util/environment.ts:9]: Fixed!

**Comment by anxolin** [comment]
Merging, feel free to review when your are back :)

**Comment by alfetopito** [comment]
Should the preview build fail?

**Comment by anxolin** [comment]
Well, this is old, so if it there was an issue I'm sure is solved by now in a follow up. I don't remember

**Pull request merged**